csharp-advanced-topics/FuncActionDelegate/Tests/FuncActionDelegateTests.cs

```diff
@@ -0,0 +1,89 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Tests
+{
+    [TestClass]
+    public class FuncActionDelegateTests
+    {
+        public class Candidate
+        {
+            public string Name { get; set; }
+            public int Age { get; set; }
+            public int Experience { get; set; }
+        }
+
+        private static bool CheckExperience(Candidate candidate)
+        {
+            return candidate.Experience >= 3;
+        }
+
+        private static bool CheckAge(Candidate candidate)
+        {
+            return candidate.Age < 40;
+        }
+
+        private static void DisplayScore(Candidate candidate, int score)
+        {
+            Console.WriteLine($"{candidate.Name}'s score is {score}");
+        }
+
+        [TestMethod]
+        public void whenActionDelegate_DelegateInvocationListNotEmpty()
+        {
+            Action<Candidate,int> executeDisplayScoreAction = DisplayScore;
+            var invocationList = executeDisplayScoreAction.GetInvocationList();
+            Assert.AreEqual(invocationList.Length,1);
+        }
+
+        [TestMethod]
+        public void whenFuncDelegate_DelegateInvocationListNotEmpty()
+        {
+            Func<Candidate, bool> checkAgeFunc = CheckAge;
+            Func<Candidate, bool> checkExperienceFunc = CheckExperience;
+            
+            var invocationList1 = checkAgeFunc.GetInvocationList();
+            var invocationList2 = checkExperienceFunc.GetInvocationList();
+
+            Assert.AreEqual(invocationList1.Length, 1);
+            Assert.AreEqual(invocationList2.Length, 1);
+        }
+
+        [TestMethod]
+        public void whenActionDelegate_DelegateExecutesTheReferenceMethod()
+        {
+            var candidate = new Candidate { Name = "David", Age = 20, Experience = 5 };
+            Action<Candidate, int> executeDisplayScoreAction= DisplayScore;
+
+            try
+            {
+                executeDisplayScoreAction(candidate, 10);
+                Assert.IsTrue(true);
+            }
+            catch
+            {
+                Assert.IsTrue(false);
+            }
```

Comment:
Not sure why you'd use try-catch here in the test. The test results should be fixed every time.

Reply:
I've used [try-catch](https://stackoverflow.com/a/37180480/579381) because it lets me know whether Action can execute the referenced method, which is a void method or not.

Reply:
One test should deal with one scenario.